Add `cargo rustdoc` for passing arbitrary flags to rustdoc

Manishearth opened the pull request:
Quite useful for doing `cargo rustdoc -- --no-defaults` to get a doc package with private items.

This is similar to `cargo rustc`, except that the flags passed to `cargo rustdoc` are applied to both the main package and dependencies.

A friend of mine was looking for this feature, couldn't find a way to do it without directly calling rustdoc. This would probably be useful for using the full configurability of rustdoc in cargo-applications.

r? @alexcrichton

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way Github handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

steveklabnik commented:
conceputally, :+1: from me.

alexcrichton commented on src/bin/rustdoc.rs:
```diff
@@ -0,0 +1,87 @@
+    -q, --quiet              No output printed to stdout
+    --color WHEN             Coloring: auto, always, never
+
+By default the documentation for the local package and all dependencies is
```
I think this may want to customize this text instead of just copy/pasting from `cargo help doc`, for example this should probably start out by saying that it is intended to configure the invocation of rustdoc on one crate and not mention much about dependencies up here at the top.

alexcrichton commented on src/bin/rustdoc.rs:
```diff
@@ -0,0 +1,87 @@
+built. The output is all placed in `target/doc` in rustdoc's usual format.
+
+The specified <opts>... will all be passed to rustdoc for the main package and
+all of its dependencies.
```
Is this true? It looks like it's only passed to the top crate in the implementation below (which is what should be done to mirror `rustc`)

alexcrichton commented on src/bin/rustdoc.rs:
```diff
@@ -0,0 +1,87 @@
+
+#[derive(RustcDecodable)]
+struct Options {
+    arg_opts: Option<Vec<String>>,
```
I think this may just be able to be `Vec<String>` (e.g. it's empty by default)

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -191,6 +195,19 @@ pub fn compile_pkg<'a>(package: &Package,
+    };
+
+    let targets = target_with_rustdoc.as_ref().map_or(targets,
+                                             |o| o.into_iter().map(|&(t, ref p)| (t, p)).collect());
```
This should have the same logic as `cargo rustc` where multiple targets are rejected (and some form of filter is needed)

alexcrichton replied:
Can you also be sure to wrap this to 80 chars?

alexcrichton replied:
(ping on the wrap)

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -191,6 +195,19 @@ pub fn compile_pkg<'a>(package: &Package,
+    let mut target_with_rustdoc = None;
+    if let Some(args) = *extra_rustdoc_args {
+        let mut target_with_rustdoc_inner = Vec::new();
+        for &(target, profile) in &targets {
```
I think you'll probably want to filter this by `profile.doc` as not all targets have rustdoc run on them by default.

alexcrichton commented on tests/test_cargo_rustdoc.rs:
```diff
@@ -0,0 +1,58 @@
+            running = RUNNING, compiling = COMPILING, sep = SEP,
+            dir = p.root().display(), url = p.url())));
+});
+
```
Can you add some test cases with dependencies as well? The arguments shouldn't be passed to upstream dependencies but the documentation for those dependencies should still be built. This should also have cases with more than one target (e.g. a lib and a binary) and an error should be generated (like with `cargo rustc`)

alexcrichton commented:
cc @brson, @wycats, a new Cargo subcommand, but it fits well within the pattern of `cargo rustc` so I'm fine landing without an RFC.

brson commented:
sgtm

Manishearth commented:
So, I actually _want_ the subcommand to pass the rustdoc flags to all dependencies. Rustdoc output would be weird if only one crate was compiled with a set of flags.

For `cargo rustc` it makes sense to limit it to the final crate, since it's mostly for tweaking linker args and whatnot. But for `cargo rustdoc` we want to finally get a uniform chunk of documentation, which means the flags should be passed down.

I thought this PR did do that, turns out it doesn't, I'll see if I can make it work so that the args are passed to upstream deps too.

alexcrichton commented:
I would personally not want to land this with subtly different behavior than `cargo rustc` when both commands seem quite similar. Even with rustdoc I can imagine there are use cases of only passing flags to the top-level crate rather than all dependencies, although I'd certainly believe that it's more rare.

Configuration of rustdoc for all crates in the dependency graph is certainly desired, but I think that wants a more principled solution than just an extra command which passes all the flags everywhere. E.g. this is a recipe for all projects having their documentation building step be something like `cargo rustdoc -- <ten obscure flags>`, where instead the options could be stored in the manifest (for example) and the idiom would still be just `cargo doc`.

Manishearth commented:
Alright, fixed so that it mirrorrs rustc

>  E.g. this is a recipe for all projects having their documentation building step be something like cargo rustdoc -- <ten obscure flags>, where instead the options could be stored in the manifest (for example) and the idiom would still be just cargo doc.


So there are multiple use cases here. One is for the default build, and one is for specialized builds. For example, `cargo build` might be the default, but I may want to occasionally be able to run `cargo rustc -- -C codegen_units=10` or something during development. Or `cargo rustc -- -Z extra-plugins=plugin -L /path/to/plugin -D foo`. These aren't part of the build step, these are tweaks I'm running during development. Editing a file seems ... unwieldy.

I'd prefer if we had both a cargo-config based solution, _and_ a terminal-based solution. The simplest one seems to be allowing `cargo rustc` and `cargo rustdoc` to have a `--flag-all` flag, which passes the provided flag to all deps (instead of the top one), and overrides any `rustc-flags`/`rustdoc-flags` in the cargo config.

alexcrichton commented:
Yes I agree that passing custom flags shouldn't always involve editing files, but I was mostly just saying that this should follow `cargo rustc`, discussion for this kind of feature probably shouldn't happen on this PR and it would likely require an RFC.

alexcrichton commented on src/bin/bench.rs:
```diff
@@ -83,6 +83,7 @@ pub fn execute(options: Options, config: &Config) -> CliResult<Option<()>> {
                                             &options.flag_test,
                                             &options.flag_example,
                                             &options.flag_bench),
+            extra_rustdoc_args: vec![],
```
Cargo tends to favor `Vec::new()` over `vec![]` currently

alexcrichton commented on src/bin/rustdoc.rs:
```diff
@@ -0,0 +1,95 @@
+    };
+
+    try!(ops::doc(&root, &mut doc_opts).map_err(|err| {
+        CliError::from_boxed(err, 101)
```
This can just be:

```rust
try!(ops::doc(...))
```

alexcrichton commented on src/cargo/ops/cargo_rustc/mod.rs:
```diff
@@ -573,9 +577,9 @@ fn build_base_args(cx: &Context,
-    } = *profile;
+        opt_level, lto, codegen_units, ref rustc_args,
+        rustdoc_args: _, debuginfo, debug_assertions, rpath,
+        test, doc: _doc } = *profile;
```
Could this keep the `} = *profile;` on its own line?

alexcrichton commented on src/bin/rustdoc.rs:
```diff
@@ -0,0 +1,95 @@
+will simply be added to the compiler invocation.
+
+This command requires that only one target is being compiled. If more than one
+target is available for the current package the filters of --lib, --bin, etc,
```
This don't seem to currently be filters, perhaps they should be added?

Manishearth replied:
Actually, this isn't needed; I think `cargo doc` only handles one lib target and we do the same.

alexcrichton replied:
I believe that we document binaries as well?

alexcrichton commented on src/bin/rustdoc.rs:
```diff
@@ -0,0 +1,95 @@
+The specified target for the current package (or package specified by SPEC if
+provided) will be compiled along with all of its dependencies. The specified
+<opts>... will all be passed to the final rustdoc invocation, not any of the
+dependencies. Note that the compiler will still unconditionally receive
```
s/compiler/rustdoc/

Manishearth commented:
(addressed)

alexcrichton commented on src/bin/rustdoc.rs:
```diff
@@ -0,0 +1,93 @@
+<opts>... will all be passed to the final rustdoc invocation, not any of the
+dependencies. Note that rustdoc will still unconditionally receive
+arguments such as -L, --extern, and --crate-type, and the specified <opts>...
+will simply be added to the compiler invocation.
```
s/compiler/rustdoc/

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -191,6 +195,27 @@ pub fn compile_pkg<'a>(package: &Package,
+            return Err(human("extra arguments to `rustdoc` can only be passed to \
+                  one target, consider filtering\nthe package by \
+                  passing e.g. `--lib` or `--bin NAME` to specify \
+                  a single target"));
```
Could you add a test for this error message? (e.g. a lib + bin in a project)

bors commented:
:umbrella: The latest upstream changes (presumably #1991) made this pull request unmergeable. Please resolve the merge conflicts.

Manishearth commented:
Updated.

It occurred to me that there's no reason for us to not allow multiple targets for `cargo rustdoc`, so I removed that restriction entirely.

alexcrichton commented:
The restriction for only one target actually mirrors `cargo rustc` to ensure that the custom flags are targeted at only one target instead of accidentally at many. I believe it's possible to have many targets in one package with e.g. many binary targets. Could the restriction be re-added?

Manishearth commented:
But why would someone care about flags applying "accidentally" to many? These are docs, not code. It makes some sense to want to avoid applying the wrong extra link args, but the rustdoc flags are pretty benign.

Besides, it's not possible to filter targets for docs in the current framework. (https://github.com/rust-lang/cargo/blob/master/src/bin/rustc.rs#L82, https://github.com/rust-lang/cargo/blob/master/src/bin/doc.rs#L69). Could be fixed though.

alexcrichton commented:
This is inconsistent with `cargo rustc`, which this is intending to mirror, so it needs to reject the situation where multiple targets may apply. Although it may be less harmful in the documentation than compile case, it's still important that these commands maintain consistency.

For filtering, that's why I was saying that the `--bin` and `--lib` flags will need to be added to `cargo rustdoc`.

bors commented:
:umbrella: The latest upstream changes (presumably #1828) made this pull request unmergeable. Please resolve the merge conflicts.

jonas-schievink commented:
@Manishearth What's the status of this? I'd love to use this for internal documentation.

Manishearth commented:
Bit busy right now. I'm not really sure how to add the other flags, and sort of disagree that we need to.

alexcrichton commented:
@Manishearth would you like me to take this over? To add the new flags you'd want to change [the compile filter](https://github.com/rust-lang/cargo/pull/1977/files#diff-3167bbbb0a228e63fb55c08ed8cf3093R75) being used and construct it in a manner similar to the other subcommands.

Manishearth commented:
Sure, take it over. I don't have a lot of continuous time to work on things, and this is at the back of my queue :)

alexcrichton commented:
Rebased in #2129
